pkg/roll/run_update.go

```diff
@@ -238,10 +230,16 @@ func (u update) lockRCs() error {
 	}
 
 	err = u.lock.Lock(oldPath)
+	if err != nil {
+		// The second key couldn't be locked, so release the first key before retrying.
+		RetryOrQuit(
+			func() error { return u.lock.Unlock(newPath) },
+			done,
+			u.logger,
+			fmt.Sprintf("unlocking %s", newPath),
+		)
+	}
 	if _, ok := err.(kp.AlreadyLockedError); ok {
```

Comment:
I will try to structure these conditionals better to make it more clear what's going on.